core/primitives-core/src/account.rs

```diff
@@ -20,6 +23,173 @@ pub struct Account {
     pub storage_usage: StorageUsage,
 }
 
+#[cfg(feature = "protocol_feature_add_account_versions")]
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct AccountV1 {
+    /// The total not locked tokens.
+    #[serde(with = "u128_dec_format_compatible")]
+    pub amount: Balance,
+    /// The amount locked due to staking.
+    #[serde(with = "u128_dec_format_compatible")]
+    pub locked: Balance,
+    /// Hash of the code stored in the storage for this account.
+    pub code_hash: CryptoHash,
+    /// Storage used by the given account, includes account id, this struct, access keys and other data.
+    pub storage_usage: StorageUsage,
+}
+
+#[cfg(feature = "protocol_feature_add_account_versions")]
+#[derive(BorshSerialize, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum Account {
+    AccountV1(AccountV1),
+}
+
+impl Account {
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn new(
+        amount: Balance,
+        locked: Balance,
+        code_hash: CryptoHash,
+        storage_usage: StorageUsage,
+    ) -> Self {
+        Account { amount, locked, code_hash, storage_usage }
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn new(
+        amount: Balance,
+        locked: Balance,
+        code_hash: CryptoHash,
+        storage_usage: StorageUsage,
+    ) -> Self {
+        Account::AccountV1(AccountV1 { amount, locked, code_hash, storage_usage })
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn amount(&self) -> Balance {
+        self.amount
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn amount(&self) -> Balance {
+        match self {
+            Account::AccountV1(acc) => acc.amount,
+        }
+    }
```

Comment:
Some Rust perf trivia: 

Rust compilation model is close enough to C++: in a compiled library, generic functions are present in the "source form" (a-la C++ templates in headers) but non-generic functions are compiled down to machine code. 

The latter means that non-generic functions can't be inlined across the crate. So, a `pub` getter without `#[inline]` annotation can really drag the performance down. Inline helps, because it instructs the compiler to include "source" of the function into the compiled library, thus making it eligible for compilation *at all*.

So, as a rule-of-thumb, small public functions should be `#[inline]`. Now, this doesn't entirely apply to us, as we compile with `lto=true` and `codegen-units=1`, which makes *all* the source code available to compiler (sort-of amalgamation build). However, this makes our compile times in release mode *really* slow, as it kills any parallelism.

Reply:
we often comment out `lto=true` when we had to debug a release build, it would be nice to have these inlines so that without lto=true version has better performance

Reply:
Done